source/common/http/http1/conn_pool.cc

```diff
@@ -209,25 +215,44 @@ void ConnPoolImpl::onResponseComplete(ActiveClient& client) {
     host_->cluster().stats().upstream_cx_max_requests_.inc();
     onDownstreamReset(client);
   } else {
-    processIdleClient(client);
+    processIdleClient(client, true);
+  }
+}
+
+void ConnPoolImpl::onUpstreamReady() {
+  upstream_ready_enabled_ = false;
+  while (!pending_requests_.empty() && !ready_clients_.empty()) {
```

Comment:
sanity check: this is simply deferring work we used to do in the last dispatcher loop to the next loop, so there's no corner case we'll end up batching together too much work and triggering the watchdog, right?

Reply:
The chance of batching together is when multiple upstream connections are completing responses at same time (same epoll callback), I don't think there will will be too much work to trigger watchdog.